File: core/parser.go
package core

import (
	"image/color"

	"github.com/valyala/fastjson"
)

func CanvasFromJSON(jsonString string) *Canvas {
	var p fastjson.Parser
	v, err := p.Parse(jsonString)
	if err != nil {
		panic(err)
	}
	return parseCanvas(v)
}

func parseRGBA(v []*fastjson.Value) *color.RGBA {
	return &color.RGBA{
		uint8(v[0].GetInt()),
		uint8(v[1].GetInt()),
		uint8(v[2].GetInt()),
		uint8(v[3].GetInt()),
	}
}

func parseCanvas(v *fastjson.Value) *Canvas {
	elements := parseElements(v.GetArray("elements"))
	return &Canvas{
		Width:           v.GetFloat64("width"),
		Height:          v.GetFloat64("height"),
		BackgroundColor: parseRGBA(v.GetArray("backgroundColor")),
		Debug:           v.GetBool("debug"),
		Elements: &CanvasElements{
			Images:    parseImages(elements["image"]),
			Textboxes: parseTextboxes(elements["textbox"]),
		},
	}
}

func parseElements(vs []*fastjson.Value) map[string][]*fastjson.Value {
	elemMap := map[string][]*fastjson.Value{}
	for _, e := range vs {
		eType := string(e.GetStringBytes("type"))
		if _, ok := elemMap[eType]; !ok {
			elemMap[eType] = []*fastjson.Value{}
		}
		elemMap[eType] = append(elemMap[eType], e)
	}
	return elemMap
}

func parseImages(vs []*fastjson.Value) []*ImageElement {
	elements := []*ImageElement{}
	for _, e := range vs {
		newElement := &ImageElement{
			Id:   e.GetInt("id"),
			Rect: parseRect(e.Get("rect")),
			Data: &ImageData{
				URI:          string(e.GetStringBytes("data", "uri")),
				BorderRadius: e.GetFloat64("data", "borderRadius"),
			},
		}
		elements = append(elements, newElement)
	}
	return elements
}

func parseTextboxes(vs []*fastjson.Value) []*TextboxElement {
	elements := []*TextboxElement{}
	for _, e := range vs {
		newElement := &TextboxElement{
			Id:   e.GetInt("id"),
			Rect: parseRect(e.Get("rect")),
			Data: &TextboxData{
				Text:            string(e.GetStringBytes("data", "text")),
				FontFamily:      string(e.GetStringBytes("data", "fontFamily")),
				FontWeight:      string(e.GetStringBytes("data", "fontWeight")),
				TextAlign:       string(e.GetStringBytes("data", "textAlign")),
				Color:           parseRGBA(e.GetArray("data", "color")),
				Caps:            e.GetBool("data", "caps"),
				OutlineWidth:    e.GetFloat64("data", "outlineWidth"),
				OutlineColor:    parseRGBA(e.GetArray("data", "outlineColor")),
				BackgroundColor: parseRGBA(e.GetArray("data", "backgroundColor")),
			},
		}
		elements = append(elements, newElement)
	}
	return elements
}

func parseRect(v *fastjson.Value) *Rect {
	return &Rect{
		X:        v.GetFloat64("x"),
		Y:        v.GetFloat64("y"),
		Width:    v.GetFloat64("width"),
		Height:   v.GetFloat64("height"),
		Rotation: v.GetFloat64("rotation"),
	}
}

File: concept/src/parser.ts
import { Canvas, CanvasElement } from "./canvas"
import { ImageElement } from "./image"
import { Rect } from "./rect"

function max(input: any, max: number) {
    return Number.isNaN(input) ? max : Math.max(input, max)
}

function parseRGBA(input: any) {
    return typeof input !== "string" ? "" : input
}

export function canvasFromJSON(json: string) {
    const obj = JSON.parse(json)
    return new ParsingContext().parseCanvas(obj)
}

class ParsingContext {
    private obj: Record<any, any> = {}
    private pr = 0

    public parseCanvas(obj: any) {
        if (typeof obj !== "object") {
            throw new Error()
        }
        this.obj = obj
        this.pr = this.obj.pixelRatio
        if (!this.pr) {
            this.pr = 1
        }
        const c = new Canvas()
        c.width = max(this.obj.width, 1) * this.pr
        c.height = max(this.obj.height, 1) * this.pr
        c.backgroundColor = parseRGBA(this.obj.backgroundColor)
        c.pixelRatio = this.pr
        c.debug = Boolean(this.obj.debug)
        c.multiPalette = Boolean(this.obj.multiPalette)
        c.elements = this.parseElements()
        c.animated = false
        return c
    }

    private parseElements() {
        if (!("elements" in this.obj) || !Array.isArray(this.obj.elements)) {
            return
        }
        const elements: CanvasElement[] = []
        this.obj.elements.map((element) => {
            if (!(typeof element === "object")) {
                return
            }
            switch (element.type) {
                case "image":
                    elements.push(this.parseImage(element))
                    break
            }
        })
        return elements
    }

    private parseImage(data: any) {
        const imageURI = data.data?.uri
        if (!imageURI) {
            throw new Error("Image URI is not defined")
        }
        const image = new ImageElement()
        image.rect = this.parseRect(data?.rect)
        image.data = {
            image: imageURI,
            borderRadius: data.borderRadius * this.pr
        }
        return image
    }

    private parseRect(data: any) {
        if (typeof data !== "object") {
            throw new Error("Rect is not defined")
        }
        const rect = new Rect()
        rect.x = data.x * this.pr
        rect.y = data.y * this.pr
        rect.width = data.width * this.pr
        rect.height = data.height * this.pr
        rect.rotation = data.rotation * this.pr
        return rect
    }
}
